engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala

```diff
@@ -0,0 +1,340 @@
+package org.enso.compiler.pass.analyse
+
+import org.enso.compiler.context.{
+  CompilerContext,
+  FramePointer,
+  InlineContext,
+  LocalScope,
+  ModuleContext
+}
+import org.enso.compiler.core.{CompilerError, IR}
+import org.enso.compiler.core.ir.expression.Application
+import org.enso.compiler.core.ir.{
+  CallArgument,
+  DefinitionArgument,
+  Expression,
+  Function,
+  Module,
+  Name,
+  ProcessingPass
+}
+import org.enso.compiler.core.Implicits.AsMetadata
+import org.enso.compiler.core.ir.MetadataStorage.MetadataPair
+import org.enso.compiler.core.ir.module.scope.Definition
+import org.enso.compiler.core.ir.module.scope.definition.Method
+import org.enso.compiler.pass.IRPass
+import org.enso.compiler.pass.IRPass.IRMetadata
+import org.enso.compiler.pass.analyse.alias.{Graph, Info}
+
+/** This pass attaches [[FramePointer]] as metadata to all the IR elements that already
+  * have [[org.enso.compiler.pass.analyse.alias.Info.Occurrence]] attached.
+  */
+case object FramePointerAnalysis extends IRPass {
+
+  override type Metadata = FramePointerMeta
+
+  override type Config = IRPass.Configuration.Default
+
+  override val precursorPasses: Seq[IRPass] = {
+    Seq(AliasAnalysis)
+  }
+
+  override val invalidatedPasses: Seq[IRPass] = Seq(this)
+
+  override def runModule(ir: Module, moduleContext: ModuleContext): Module = {
+    val newBindings = ir.bindings.map(processBinding)
+    ir.copy(bindings = newBindings)
+  }
+
+  private def processBinding(
+    ir: Definition
+  ): Definition = {
+    ir match {
+      case m: Method.Explicit =>
+        getAliasAnalysisGraph(m) match {
+          case Some(graph) =>
+            m.copy(
+              body = processExpression(m.body, graph)
+            )
+          case _ => m
+        }
+      case m: Method.Conversion =>
+        getAliasAnalysisGraph(m) match {
+          case Some(graph) =>
+            m.copy(
+              body = m.body.mapExpressions(processExpression(_, graph))
+            )
+          case _ => m
+        }
+      case t: Definition.Type =>
+        getAliasAnalysisGraph(t) match {
+          case Some(graph) =>
+            t.copy(
```

Comment:
- btw. see @hubertp #10718 to see most recent thoughts about minimizing copying

Reply:
Copying removed in 61821bf3932f72bac1355f2aed3c05af6972d02b